### DISCUSSION TITLE
Intel(R) Xeon(R) Gold 6248R CPU is optimized for maximum performance?

### DISCUSSION THREAD

**Discussion by Xie-ry** [OpenMathLib/OpenBLAS#4448] [General]
The reason was that after replacing mkl with openblas, I found that the program was slowing down.
Then I checked the openblas compilation, and I found that openblas v0.3.26 was compiled by default without using avx512.
`make CC=gcc FC=gfortran  BINARY=64 NO_STATIC=1 BIGNUMA=0`
`-DVERSION=\"0.3.26\" -msse3 -mssse3 -msse4.1 -mavx -UASMNAME -UASMFNAME -UNAME -UCNAME ,,,`

lscpu show me that 6248R support :
` avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl ,,,`

If I explicitly turn on avx compilation, I have to add some extra compilation flags to make the compilation successful. 
`
make CC=gcc FC=gfortran  BINARY=64 NO_STATIC=1 BIGNUMA=0 NO_AVX512=0 COMMON_OPT="-O2 -mavx512f -mavx512dq -mavx512vl -mfma"
`
However, this results in slightly slower performance than the default compilation.

from wiki, i see :
>  In Intel's process–architecture–optimization model, Cascade Lake is an optimization of Skylake.

I don't see Cascade Lake in the target list, so I guess openblas has not maximized the optimization of this chip?

**Comment by martin-frbg** [comment]
Yes, Cascade Lake is supported by the SKYLAKEX target and not all BLAS kernels for it are available in AVX512 versions yet. However, if the default compilation on your system resulted in only avx/avx2 flags showing up in the compile commands, your compiler may not be new enough to support AVX512 instructions (in that case, the c_check script automatically adds NO_AVX512=1 to the options for make). Which compiler are you using ?

**Reply by Xie-ry** [reply]
gcc 9.1
`Target: x86_64-pc-linux-gnu
Configured with: ../gcc-9.1.0/configure CC=/usr/bin/gcc CXX=/usr/bin/g++ --enable-threads=posix --enable--long-long --enable-languages=c,c++,fortran --disable-multilib
Thread model: posix
gcc version 9.1.0 (GCC)`

**Reply by Xie-ry** [reply]
Sorry, I wrote a wrong place : "If I explicitly turn off avx compilation".  Actually should be "turn on".

**Comment by martin-frbg** [comment]
Hmm, that's strange, 9.1 should support AVX512 including the corresponding C intrinsics (immintrin.h) that c_check is looking for. So you should be getting an AVX512 build (optimized for Skylake Xeon, but the differences should be marginal) by default. Maybe your assembler (binutils package) is too old for AVX512 - could you check if the Makefile.conf generated during a default build has "NO_AVX512=1" in it ?

**Reply by Xie-ry** [reply]
By check Makefile.conf, I see HAVE_AVX512VL=1.
In fact, the two different writing methods generate the same Makefile.conf file.

`
make CC=gcc FC=gfortran  BINARY=64 NO_STATIC=1    # make 1
`

`
make CC=gcc FC=gfortran  BINARY=64 NO_STATIC=1 BIGNUMA=0 NO_AVX512=0 COMMON_OPT="-O2 -mavx512f -mavx512dq -mavx512vl -mfma"    # make 2
`

I tested matrix multiplication and SVD using demo. The result shows that the performance of "make 1" is the same as that of mkl (mat = 50 * 50, about 0.005ms).
But the "make 2", nedd 0.12ms.

**Reply by martin-frbg** [reply]
maybe there is still a spot in the code where it just checks if NO_AVX512 is defined at all, and not if it is set to 1 - this used to be a problem with many of the NO_something options in the past, as it was assumed that they would be specified only when set to 1

**Comment by martin-frbg** [comment]
But if you are comparing against MKL, if your code uses many LAPACK functions instead of direct BLAS calls, it could well be that you are getting optimized/parallelized LAPACK functions from MKL too, while most of the LAPACK in OpenBLAS is copied from the unoptimized Reference-LAPACK ("netlib") implementation.

**Reply by Xie-ry** [reply]
Maybe my benchmarks were wrong, but now I see that simple multiplication openblas is as fast as mkl.
I have found a new problem and need your help. I open a new discussion about the dgemm_ function.
